### PR TITLE
Improvements in String.toLowerCase

### PR DESCRIPTION
**Pull request by romainbrenguier**
This extends add_axioms_for_to_lower_case by taking into account latin
characters up to u00FF and adding in the lemmas the property that the
characters should not exceed u00FF, making the axioms correct (we will
not generate wrong assertion violation) but incomplete (we may miss
cases involving characters beyond u00FF).
This also adds more checks to the `java_case` regression test.

### CODE REVIEW

File: src/solvers/refinement/string_constraint_generator_transformation.cpp [hunk 1]
```diff
@@ -216,34 +216,48 @@ string_exprt string_constraint_generatort::add_axioms_for_to_lower_case(
...
   symbol_exprt idx=fresh_univ_index("QA_lower_case", index_type);
-  exprt is_upper_case=and_exprt(
+  exprt::operandst upper_case;
+  upper_case.push_back(and_exprt(
     binary_relation_exprt(char_A, ID_le, str[idx]),
-    binary_relation_exprt(str[idx], ID_le, char_Z));
-  minus_exprt diff(char_a, char_A);
-  equal_exprt convert(res[idx], plus_exprt(str[idx], diff));
-  string_constraintt a2(idx, res.length(), is_upper_case, convert);
+    binary_relation_exprt(str[idx], ID_le, from_integer('Z', char_type))));
```
Comment (mgudemann) [inline src/solvers/refinement/string_constraint_generator_transformation.cpp]: please also define `char_Z`, too

File: src/solvers/refinement/string_constraint_generator_transformation.cpp [hunk 2]
```diff
@@ -216,34 +216,48 @@ string_exprt string_constraint_generatort::add_axioms_for_to_lower_case(
...
+
+  upper_case.push_back(and_exprt(
+    binary_relation_exprt(from_integer(0xc0, char_type), ID_le, str[idx]),
```
Comment (mgudemann) [inline src/solvers/refinement/string_constraint_generator_transformation.cpp]: please add a comment for what `0xc0` is

File: src/solvers/refinement/string_constraint_generator_transformation.cpp [hunk 3]
```diff
@@ -216,34 +216,48 @@ string_exprt string_constraint_generatort::add_axioms_for_to_lower_case(
...
+    binary_relation_exprt(str[idx], ID_le, from_integer(0xd6, char_type))));
+
+  upper_case.push_back(and_exprt(
+    binary_relation_exprt(from_integer(0xd8, char_type), ID_le, str[idx]),
```
Comment (mgudemann) [inline src/solvers/refinement/string_constraint_generator_transformation.cpp]: please add a comment for what `0xd8` is

File: src/solvers/refinement/string_constraint_generator_transformation.cpp [hunk 4]
```diff
@@ -216,34 +216,48 @@ string_exprt string_constraint_generatort::add_axioms_for_to_lower_case(
   string_exprt res=fresh_string(ref_type);
   exprt char_a=constant_char('a', char_type);
   exprt char_A=constant_char('A', char_type);
```
Comment (mgudemann) [inline src/solvers/refinement/string_constraint_generator_transformation.cpp]: shouldn't these be `const` ?

File: src/solvers/refinement/string_constraint_generator_transformation.cpp [hunk 5]
```diff
@@ -216,34 +216,48 @@ string_exprt string_constraint_generatort::add_axioms_for_to_lower_case(
...
 
   // We add axioms:
   // a1 : |res| = |str|
-  // a2 : forall idx<str.length, 'A'<=str[idx]<='Z' => res[idx]=str[idx]+'a'-'A'
-  // a3 : forall idx<str.length, !('a'<=str[idx]<='z') => res[idx]=str[idx]
-  // forall idx<str.length,
-  // this[idx]='A'<=str[idx]<='Z' ? str[idx]+'a'-'A' : str[idx]
+  // a2 : forall idx<str.length,
+  //   is_upper_case(str[idx])?
+  //      res[idx]=str[idx]+'a'-'A' : res[idx]=str[idx]<0x100
```
Comment (mgudemann) [inline src/solvers/refinement/string_constraint_generator_transformation.cpp]: the actual implementation is slightly different, its `res[idx]=str[idx]+0x20`, maybe replace by `diff` here and add `where diff='a'-'A'=0x20`

File: src/solvers/refinement/string_constraint_generator_transformation.cpp (hunk 3 above)
Comment (allredj) [inline src/solvers/refinement/string_constraint_generator_transformation.cpp]: Ideally, name variables to hold these values, removing the need for comments.

**Pull request merged**